urlfcc351a1ba_kuntapavankumar_Z_tgz-pJ8-edu_concordia_dpis_ReplicaTestJ8/src/edu/concordia/dpis/stationserver/FaultyStationServerImpl.java

```diff
@@ -0,0 +1,497 @@
+	}
+
+	// make sure the list is ready to insert new record
+	private void newRecordsList(final String lastName) {
+		final List<Record> records = this.records.get(Character
+				.toUpperCase(lastName.charAt(0)));
+
+		if (records == null) {
+			synchronized (this.records) {
+				this.records.put(Character.toUpperCase(lastName.charAt(0)),
+						new ArrayList<Record>());
+			}
+		}
+	}
+
+	// check for last name null and emptiness
+	private boolean hasLastName(String lastName) {
+		return lastName != null && !lastName.isEmpty();
+	}
+
+	// new record id is of length 5 character and is in a running sequence
+	private String newRecordId(RecordType type) {
+		String recordId;
+
+		synchronized (this.records) {
+			// boundary case, when running out of id's
+			if (records.size() > 99999) {
+				throw new RuntimeException("Out of primary key for record");
+			}
+			recordId = type.getCode()
+					+ String.format("%05d", (countRecords() + 1));
+		}
+		return recordId;
+	}
+
+	// make an entry into the records
+	private void insertNewRecord(String lastName, Record record) {
+		List<Record> recordsList = this.records.get(Character
+				.toUpperCase(lastName.charAt(0)));
+		synchronized (recordsList) {
+			this.records.get(Character.toUpperCase(lastName.charAt(0))).add(
+					record);
+		}
+	}
+
+	// brings the count of all the stations record count
+	public String getRecordCounts(String badgeId) {
+		return "";
+	}
+
+	// move a record completely from one station to the other
+	public boolean transferRecord(String badgeId, String recordId,
+			String remoteStationServerName) {
+		return false;
+	}
+
+	// return ths number of records in this stations
+	private long countRecords() {
+		long no_of_records = 0;
+		for (Character ch : records.keySet()) {
+			no_of_records += this.records.get(ch).size();
+		}
+		return no_of_records;
+	}
+
+	// a TCP Client program to invoke the TCP server on the other side
+	class TCPClient {
+		private boolean transfer(String record, StationType station) {
+			Socket s = null;
+
+			String[] hostAndPort = extractHostAndPortNumber((String) tcpProperties
+					.get(station.getStationCode().toString()));
+			int serverPort = Integer.parseInt(hostAndPort[1]);
+			try {
+				s = new Socket(hostAndPort[0], serverPort);
+				// input stream
+				DataInputStream in = new DataInputStream(s.getInputStream());
+				// output stream
+				DataOutputStream out = new DataOutputStream(s.getOutputStream());
+
+				out.writeUTF(record);
+				boolean result = Boolean.parseBoolean(in.readUTF());
+				return result;
+			} catch (IOException e) {
+				e.printStackTrace();
+				log.error(e.getMessage());
+			}
+			return false;
+		}
+	}
+
+	// A UDPClient to invoke the UDPServer
+	class UDPClient {
+
+		private String ping(String remoteHost, int serverPort)
+				throws FileNotFoundException, IOException {
+			String response = "";
+			Properties properties = new Properties();
+			properties.load(new FileInputStream("udp.properties"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileInputStream*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *FileInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).